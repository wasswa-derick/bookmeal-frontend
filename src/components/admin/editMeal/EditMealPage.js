import React from "react";
import { connect } from "react-redux";
import Loader from "react-loader";
import validator from "validator";
import PropTypes from "prop-types";
import { FormInput } from "../../common/forms";
import { InlineError } from "../../common";
import { getMeal, editMeal } from "../../../actions/meals";

/**
 * @export
 * @class EditMealPage
 * @extends {React.Component}
 */
export class EditMealPage extends React.Component {
  state = {
    data: {
      title: "",
      price: "",
      description: ""
    },
    errors: {},
    loaded: true
  };

  componentWillMount = () => {
    const { id } = this.props.match.params;
    this.props
      .getMeal(id)
      .then(() => {
        const { meal } = this.props;
        this.setState({ data: { ...meal } });
      })
      .catch(() => {});
  };

  /**
   * @param {Event} e
   * @returns {null} sets state on target value on change
   * @memberof EditMealPage
   */
  onChange = e =>
    this.setState({
      data: { ...this.state.data, [e.target.name]: e.target.value }
    });

  onSubmit = () => {
    const { data } = this.state;
    let errors = this.validate(data);
    if (Object.keys(errors).length === 0) {
      this.props
        .editMeal({ ...data, price: parseInt(data.price, 10) })
        .then(() => {
          this.props.history.push("/admin/meals");
        })
        .catch(err => {
          switch (err.response.status) {
            case 400:
              errors = { ...err.response.data.errors };
              this.setState({ errors });
              break;
            default:
              break;
          }
        });
    } else {
      this.setState({ errors });
    }
  };

  /**
   * @param {Object} data
   *@returns {Object} of errors
   */
  validate = data => {
    const errors = [];
    if (validator.isEmpty(String(data.price))) {
      errors.price = "This field is required";
    } else if (!validator.isNumeric(String(data.price))) {
      errors.price = "Price value should be a number";
    } else if (data.price <= 0) {
      errors.price = "Price value cannot be less or equal to zero";
    }
    if (validator.isEmpty(data.title)) {
      errors.title = "This field is required";
    }
    if (validator.isEmpty(data.description)) {
      errors.description = "This field is required";
    }
    return errors;
  };

  /**
   *
   * @returns {null} elements to render
   * @memberof EditMealPage
   */
  render() {
    const { errors, data, loaded } = this.state;
    return (
      <div className="container">
        <h5>Edit Meal: {data.title}</h5>
        <Loader loaded={loaded}>
          <div className="modal-dialog" role="document">
            <div className="modal-content">
              <div className="modal-header" />
              <div className="modal-body">
                <form>
                  <FormInput
                    name="title"
                    type="text"
                    label="Title"
                    value={data.title}
                    error={errors.title}
                    onChange={this.onChange}
                  />
                  <FormInput
                    name="price"
                    type="text"
                    label="Price"
                    value={String(data.price)}
                    onChange={this.onChange}
                    error={errors.price}
                  />
                  <div className="form-group">
                    <label htmlFor="desc">Description</label>
                    <textarea
                      className={
                        errors.description
                          ? "form-control is-invalid"
                          : "form-control"
                      }
                      name="description"
                      id="desc"
                      cols="20"
                      rows="5"
                      onChange={this.onChange}
                      value={data.description}
                    />
                    {errors.description && (
                      <InlineError text={errors.description} />
                    )}
                  </div>

                  <input className="form-control" type="file" name="image" />
                </form>
              </div>
              <div className="modal-footer">
                <button
                  type="button"
                  className="btn btn-secondary"
                  data-dismiss="modal"
                >
                  Close
                </button>
                <button
                  type="button"
                  onClick={this.onSubmit}
                  className="btn-save btn btn-primary"
                >
                  Save
                </button>
              </div>
            </div>
          </div>
        </Loader>
      </div>
    );
  }
}

EditMealPage.propTypes = {
  getMeal: PropTypes.func.isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string.isRequired
    })
  }).isRequired,
  meal: PropTypes.shape({
    id: PropTypes.number.isRequired,
    price: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired
  }).isRequired,
  editMeal: PropTypes.func.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired
};

const mapStateToProps = state => ({
  meal: state.mealsReducer.meal
});

export default connect(mapStateToProps, { getMeal, editMeal })(EditMealPage);
