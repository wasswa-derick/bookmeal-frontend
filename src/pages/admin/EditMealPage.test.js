import React from "react";
import { shallow } from "enzyme";
import { EditMealPage } from "./EditMealPage";

const err = {
  response: {
    status: 400
  }
};
const fn = () => Promise.reject(err);

describe("EditMealPage", () => {
  let wrapper;
  beforeEach(() => {
    const meal = {
      id: 0,
      title: "",
      description: "",
      price: 0
    };
    const history = {
      pathname: "",
      push: jest.fn
    };
    const match = { params: { id: "1" } };

    wrapper = shallow(
      <EditMealPage
        meal={meal}
        getMeal={fn}
        match={match}
        editMeal={fn}
        history={history}
      />
    );
  });

  it("should render correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
