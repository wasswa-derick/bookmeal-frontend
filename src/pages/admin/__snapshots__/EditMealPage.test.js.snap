// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditMealPage should render correctly 1`] = `
<div
  className="container"
>
  <h5>
    Edit Meal: 
  </h5>
  <Component
    component="div"
    loaded={true}
    loadedClassName="loadedContent"
    parentClassName="loader"
  >
    <div
      className="modal-dialog"
      role="document"
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-header"
        />
        <div
          className="modal-body"
        >
          <form>
            <FormInput
              label="Title"
              name="title"
              onChange={[Function]}
              type="text"
              value=""
            />
            <FormInput
              label="Price"
              name="price"
              onChange={[Function]}
              type="text"
              value=""
            />
            <div
              className="form-group"
            >
              <label
                htmlFor="desc"
              >
                Description
              </label>
              <textarea
                className="form-control"
                cols="20"
                id="desc"
                name="description"
                onChange={[Function]}
                rows="5"
                value=""
              />
            </div>
            <input
              className="form-control"
              name="image"
              type="file"
            />
          </form>
        </div>
        <div
          className="modal-footer"
        >
          <button
            className="btn btn-secondary"
            data-dismiss="modal"
            type="button"
          >
            Close
          </button>
          <button
            className="btn-save btn btn-primary"
            onClick={[Function]}
            type="button"
          >
            Save
          </button>
        </div>
      </div>
    </div>
  </Component>
</div>
`;
